# ECON3203 Group Project: Sensational

## Explore & Prepare data

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np

In [2]:
# load data
sample_url="https://drive.google.com/uc?export=download&id=1ekCjJOTjGD0OASErwVnwhJuPD0hFfCEF"
out_url=sample_url
db_sample = pd.read_csv(sample_url)
db_out = pd.read_csv(out_url)

In [3]:
db_sample.head()

,score_read,score_math,gender,birth,lunch,ethnicity,class_type,school,degree,ladder,experience,t_ethnicity,schooldistrict_id,school_id
0,474.0,602.0,male,1980 Q2,free,cauc,small,rural,master,level1,6.0,cauc,36.0,72.0
1,403.0,434.0,male,1980 Q3,free,afam,small,rural,bachelor,apprentice,1.0,cauc,24.0,57.0
2,455.0,520.0,male,1980 Q2,non-free,cauc,regular+aide,suburban,bachelor,level1,5.0,afam,22.0,51.0
3,405.0,454.0,female,1979 Q4,free,cauc,regular+aide,rural,bachelor,level1,11.0,cauc,4.0,6.0
4,433.0,473.0,female,1980 Q3,non-free,cauc,regular,rural,bachelor,level1,2.0,cauc,40.0,76.0


- count → number of valid (non-missing) values;
- unique → number of distinct categories (categorical data only);
- top → modal (most common) category;
- freq → how often that top category appears.


In [4]:
# Simple data stats
db_sample.describe(include="all")

,score_read,score_math,gender,birth,lunch,ethnicity,class_type,school,degree,ladder,experience,t_ethnicity,schooldistrict_id,school_id
count,4643.000000,4706.000000,5060,5055,5039,5059,5060,5060,5041,4549,5041.000000,5008,5060.000000,5060.000000
unique,NaN,NaN,2,15,2,6,3,4,4,5,NaN,2,NaN,NaN
top,NaN,NaN,male,1980 Q3,non-free,cauc,regular+aide,rural,bachelor,level1,NaN,cauc,NaN,NaN
freq,NaN,NaN,2591,1333,2578,3389,1786,2351,3306,3737,NaN,4193,NaN,NaN
mean,436.272238,485.123459,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,9.277723,NaN,18.699209,40.042885
std,31.358938,47.674836,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,5.854946,NaN,11.046311,22.770969
min,346.000000,288.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.000000,NaN,1.000000,1.000000
25%,414.000000,454.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,5.000000,NaN,11.000000,21.000000
50%,433.000000,478.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,9.000000,NaN,15.000000,39.000000
75%,453.000000,513.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,13.000000,NaN,27.000000,60.000000


In [5]:
# Categorical vars exploration
categorical_cols = db_sample.select_dtypes(include='object').columns

for col in categorical_cols:
    print(db_sample[col].value_counts(),"\n")


gender
male      2591
female    2469
Name: count, dtype: int64 

birth
1980 Q3    1333
1980 Q1    1232
1980 Q2    1222
1979 Q4     901
1979 Q3     160
1979 Q2      69
1980 Q4      53
1979 Q1      43
1978 Q4      29
1978 Q2       4
1981 Q1       2
1978 Q3       2
1981 Q4       2
1978 Q1       2
1981 Q3       1
Name: count, dtype: int64 

lunch
non-free    2578
free        2461
Name: count, dtype: int64 

ethnicity
cauc        3389
afam        1646
asian         11
other          6
hispanic       5
amindian       2
Name: count, dtype: int64 

class_type
regular+aide    1786
regular         1761
small           1513
Name: count, dtype: int64 

school
rural         2351
inner-city    1134
suburban      1123
urban          452
Name: count, dtype: int64 

degree
bachelor      3306
master        1581
master+        121
specialist      33
Name: count, dtype: int64 

ladder
level1        3737
apprentice     413
probation      261
level2          97
level3          41
Name: count, dtype: int64 


Explore Missing data if it is randomly or not

In [6]:
# Count missing values and calculate missing rate per column
missing_counts = db_sample.isnull().sum()
missing_rate = db_sample.isnull().mean()

# Combine into a summary DataFrame
missing_summary = pd.DataFrame({
    'Missing Count': missing_counts,
    'Missing Rate (%)': (missing_rate * 100).round(2)
})
# Count rows with at least one missing value
num_missing_rows = db_sample.isnull().any(axis=1).sum()

# Display
print(f"Missing Value Summary: {num_missing_rows} ({(num_missing_rows/len(db_sample)*100).round(2)}%) unique missing rows")
missing_summary

Missing Value Summary: 903 (17.85%) unique missing rows


,Missing Count,Missing Rate (%)
score_read,417,8.24
score_math,354,7.00
gender,0,0.00
birth,5,0.10
lunch,21,0.42
ethnicity,1,0.02
class_type,0,0.00
school,0,0.00
degree,19,0.38
ladder,511,10.10


In [7]:
def missing_rate(df, var, group_by):
    """
    Prints missingness rate of a variable grouped by another variable.

    Parameters:
    - df: pandas DataFrame
    - var: column name to check for missing values (string)
    - group_by: column name to group by (string)
    """
    # Calculate missingness rate
    missing_rate = df[var].isna().groupby(df[group_by]).mean()

    # Filter to show only groups with missing values
    missing_groups = missing_rate[missing_rate > 0]

    # Print result
    print(f"\nMissing rate for '{var}' grouped by '{group_by}': {len(missing_groups)} missing")
    print(missing_groups.round(4))

missing_rate(db_sample, "score_read", "gender")
missing_rate(db_sample, "ladder", "school")
missing_rate(db_sample, "ladder", "school_id")
missing_rate(db_sample, "ladder", "schooldistrict_id")



Missing rate for 'score_read' grouped by 'gender': 2 missing
gender
female    0.0826
male      0.0822
Name: score_read, dtype: float64

Missing rate for 'ladder' grouped by 'school': 4 missing
school
inner-city    0.1323
rural         0.0515
suburban      0.1781
urban         0.0885
Name: ladder, dtype: float64

Missing rate for 'ladder' grouped by 'school_id': 24 missing
school_id
1.0     0.5424
4.0     0.2353
17.0    0.1875
18.0    0.2817
20.0    0.5246
25.0    0.7407
27.0    0.1707
28.0    0.2833
32.0    0.5579
33.0    0.0952
34.0    0.2321
35.0    0.3704
37.0    0.2593
42.0    0.2295
43.0    0.4746
44.0    0.4789
45.0    0.3111
47.0    0.2245
48.0    0.2317
50.0    0.2059
51.0    0.1121
58.0    0.2051
71.0    0.3333
73.0    0.1754
Name: ladder, dtype: float64

Missing rate for 'ladder' grouped by 'schooldistrict_id': 13 missing
schooldistrict_id
1.0     0.5424
4.0     0.0870
11.0    0.1544
12.0    0.1953
13.0    0.2593
18.0    0.3147
19.0    0.2245
20.0    0.2317
21.0    0.1228
22

- 19 rows of missing experience is not random, only in school_id = 48 or school_district = 20 missing 23% of this variable among their groups. Those schools are in urban (4% missing), class_type is regular_aide (1%), ethnicity is cauc (0.56%).

-> Drop 19 missing rows of ladder
- 511 rows of missing ladder potentially not random, 54% of the missing ladder is in school_id=1 or school_district=1.

-> Label the missing rows in ladder and clasified as 'Missing'

In [8]:
# Filter and display all columns for school_id = 1
school_1_data = db_sample[db_sample['school_id'] == 1]

print(school_1_data.count())

score_read           53
score_math           53
gender               59
birth                59
lunch                59
ethnicity            59
class_type           59
school               59
degree               59
ladder               27
experience           59
t_ethnicity          59
schooldistrict_id    59
school_id            59
dtype: int64


In [9]:
# Drop missing rows in experience
db_sample.dropna(subset=["experience"], inplace=True)

# Fill in missing values in ladder with new label 'Missing'
db_sample["ladder"] = db_sample["ladder"].fillna("Missing")

# Count value occurrences in the 'ladder' column
print(db_sample['ladder'].value_counts())
db_sample.describe(include='all')

ladder
level1        3737
Missing        492
apprentice     413
probation      261
level2          97
level3          41
Name: count, dtype: int64


,score_read,score_math,gender,birth,lunch,ethnicity,class_type,school,degree,ladder,experience,t_ethnicity,schooldistrict_id,school_id
count,4624.000000,4687.000000,5041,5036,5020,5040,5041,5041,5041,5041,5041.000000,5008,5041.000000,5041.000000
unique,NaN,NaN,2,15,2,6,3,4,4,6,NaN,2,NaN,NaN
top,NaN,NaN,male,1980 Q3,non-free,cauc,regular+aide,rural,bachelor,level1,NaN,cauc,NaN,NaN
freq,NaN,NaN,2583,1330,2569,3370,1767,2351,3306,3737,NaN,4193,NaN,NaN
mean,436.298443,485.183913,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,9.277723,NaN,18.694307,40.012894
std,31.396538,47.722186,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,5.854946,NaN,11.066824,22.808599
min,346.000000,288.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.000000,NaN,1.000000,1.000000
25%,414.000000,454.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,5.000000,NaN,11.000000,21.000000
50%,433.000000,478.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,9.000000,NaN,15.000000,39.000000
75%,453.000000,513.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,13.000000,NaN,27.000000,60.000000


## Modeling

In [10]:
# Complicate name in birth , just gonna remove for now
db_sample = db_sample.drop(columns=['birth'])

In [11]:
# Drop missing values of score_read
db_read = db_sample.dropna(subset=["score_read"])
db_read.describe(include='all')

,score_read,score_math,gender,lunch,ethnicity,class_type,school,degree,ladder,experience,t_ethnicity,schooldistrict_id,school_id
count,4624.000000,4621.000000,4624,4609,4624,4624,4624,4624,4624,4624.000000,4593,4624.000000,4624.000000
unique,NaN,NaN,2,2,6,3,4,4,6,NaN,2,NaN,NaN
top,NaN,NaN,male,non-free,cauc,regular+aide,rural,bachelor,level1,NaN,cauc,NaN,NaN
freq,NaN,NaN,2370,2367,3111,1620,2201,3007,3459,NaN,3875,NaN,NaN
mean,436.298443,485.435187,NaN,NaN,NaN,NaN,NaN,NaN,NaN,9.326990,NaN,18.745242,40.023356
std,31.396538,47.665392,NaN,NaN,NaN,NaN,NaN,NaN,NaN,5.819284,NaN,11.146944,23.016896
min,346.000000,339.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,0.000000,NaN,1.000000,1.000000
25%,414.000000,454.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,5.000000,NaN,11.000000,20.000000
50%,433.000000,484.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,9.000000,NaN,15.000000,39.000000
75%,453.000000,513.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,13.000000,NaN,28.000000,61.000000


In [12]:
# Make catagorical vars into dummy vars
db_read=pd.get_dummies(db_read, drop_first=True)
db_read.rename(columns={'lunch_non-free': 'lunch_nonfree'}, inplace=True)
db_read.rename(columns={'class_type_regular+aide': 'class_type_regular_aide'}, inplace=True)
db_read.rename(columns={'degree_master+': 'degree_masterplus'}, inplace=True)
db_read.describe(include="all")

,score_read,score_math,experience,schooldistrict_id,school_id,gender_male,lunch_nonfree,ethnicity_amindian,ethnicity_asian,ethnicity_cauc,...,school_urban,degree_master,degree_masterplus,degree_specialist,ladder_apprentice,ladder_level1,ladder_level2,ladder_level3,ladder_probation,t_ethnicity_cauc
count,4624.000000,4621.000000,4624.000000,4624.000000,4624.000000,4624,4624,4624,4624,4624,...,4624,4624,4624,4624,4624,4624,4624,4624,4624,4624
unique,NaN,NaN,NaN,NaN,NaN,2,2,2,2,2,...,2,2,2,2,2,2,2,2,2,2
top,NaN,NaN,NaN,NaN,NaN,True,True,False,False,True,...,False,False,False,False,False,True,False,False,False,True
freq,NaN,NaN,NaN,NaN,NaN,2370,2367,4622,4613,3111,...,4225,3149,4512,4594,4241,3459,4534,4583,4405,3875
mean,436.298443,485.435187,9.326990,18.745242,40.023356,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
std,31.396538,47.665392,5.819284,11.146944,23.016896,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
min,346.000000,339.000000,0.000000,1.000000,1.000000,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
25%,414.000000,454.000000,5.000000,11.000000,20.000000,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
50%,433.000000,484.000000,9.000000,15.000000,39.000000,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
75%,453.000000,513.000000,13.000000,28.000000,61.000000,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [ ]:
db_sample

# need to filter some variables.



,score_read,score_math,gender,lunch,ethnicity,class_type,school,degree,ladder,experience,t_ethnicity,schooldistrict_id,school_id
0,474.0,602.0,male,free,cauc,small,rural,master,level1,6.0,cauc,36.0,72.0
1,403.0,434.0,male,free,afam,small,rural,bachelor,apprentice,1.0,cauc,24.0,57.0
2,455.0,520.0,male,non-free,cauc,regular+aide,suburban,bachelor,level1,5.0,afam,22.0,51.0
3,405.0,454.0,female,free,cauc,regular+aide,rural,bachelor,level1,11.0,cauc,4.0,6.0
4,433.0,473.0,female,non-free,cauc,regular,rural,bachelor,level1,2.0,cauc,40.0,76.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...
5055,434.0,520.0,female,free,cauc,small,rural,master,level1,8.0,cauc,40.0,76.0
5056,470.0,576.0,male,non-free,cauc,small,urban,master,level1,17.0,cauc,20.0,48.0
5057,474.0,626.0,female,non-free,cauc,regular,rural,master,level1,15.0,cauc,5.0,7.0
5058,NaN,NaN,female,free,afam,small,inner-city,bachelor,level1,11.0,cauc,11.0,27.0


### OLS of score_read

In [13]:
import statsmodels.api as sm
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from statsmodels.formula.api import ols

# Define X and y
X = db_read.drop(columns=['score_read', 'score_math'])
y = db_read['score_read']

formula = f'score_read ~ {" + ".join(X.columns)}'
print(formula)

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Fit OLS model full
OLSmodel = ols(formula, data=pd.concat([X_train, y_train], axis =1)).fit()

# Predict and evaluate
y_pred = OLSmodel.predict(X_test)
mse = mean_squared_error(y_test, y_pred)

# Print summary and MSE
print(OLSmodel.summary())
print(f"Mean Squared Error: {mse}")

score_read ~ experience + schooldistrict_id + school_id + gender_male + lunch_nonfree + ethnicity_amindian + ethnicity_asian + ethnicity_cauc + ethnicity_hispanic + ethnicity_other + class_type_regular_aide + class_type_small + school_rural + school_suburban + school_urban + degree_master + degree_masterplus + degree_specialist + ladder_apprentice + ladder_level1 + ladder_level2 + ladder_level3 + ladder_probation + t_ethnicity_cauc
                            OLS Regression Results                            
Dep. Variable:             score_read   R-squared:                       0.116
Model:                            OLS   Adj. R-squared:                  0.110
Method:                 Least Squares   F-statistic:                     20.08
Date:                Mon, 27 Oct 2025   Prob (F-statistic):           5.30e-81
Time:                        00:56:25   Log-Likelihood:                -17772.
No. Observations:                3699   AIC:                         3.559e+04
Df Residual

### OLS for score_math

In [14]:
# Drop missing values of score_math
db_math = db_sample.dropna(subset=["score_math"])

# MAke dammu vars
db_math=pd.get_dummies(db_math, drop_first=True)
db_math.rename(columns={'lunch_non-free': 'lunch_nonfree'}, inplace=True)
db_math.rename(columns={'class_type_regular+aide': 'class_type_regular_aide'}, inplace=True)
db_math.rename(columns={'degree_master+': 'degree_masterplus'}, inplace=True)
db_math.describe(include="all")

,score_read,score_math,experience,schooldistrict_id,school_id,gender_male,lunch_nonfree,ethnicity_amindian,ethnicity_asian,ethnicity_cauc,...,school_urban,degree_master,degree_masterplus,degree_specialist,ladder_apprentice,ladder_level1,ladder_level2,ladder_level3,ladder_probation,t_ethnicity_cauc
count,4621.000000,4687.000000,4687.000000,4687.000000,4687.000000,4687,4687,4687,4687,4687,...,4687,4687,4687,4687,4687,4687,4687,4687,4687,4687
unique,NaN,NaN,NaN,NaN,NaN,2,2,2,2,2,...,2,2,2,2,2,2,2,2,2,2
top,NaN,NaN,NaN,NaN,NaN,True,True,False,False,True,...,False,False,False,False,False,True,False,False,False,True
freq,NaN,NaN,NaN,NaN,NaN,2402,2403,4685,4676,3144,...,4283,3202,4573,4657,4296,3489,4597,4646,4458,3924
mean,436.303614,485.183913,9.301899,18.729038,40.025603,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
std,31.405520,47.722186,5.849657,11.105611,22.923302,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
min,346.000000,288.000000,0.000000,1.000000,1.000000,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
25%,414.000000,454.000000,5.000000,11.000000,21.000000,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
50%,433.000000,478.000000,9.000000,15.000000,39.000000,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
75%,453.000000,513.000000,13.000000,27.000000,60.000000,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [15]:
# Define X and y
X = db_math.drop(columns=['score_read', 'score_math'])
y = db_math['score_math']

formula = f'score_math ~ {" + ".join(X.columns)}'
print(formula)

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Fit OLS model full
OLSmodel = ols(formula, data=pd.concat([X_train, y_train], axis=1)).fit()

# Predict and evaluate
y_pred = OLSmodel.predict(X_test)
mse = mean_squared_error(y_test, y_pred)

# Print summary and MSE
print(OLSmodel.summary())
print(f"Mean Squared Error: {mse}")

score_math ~ experience + schooldistrict_id + school_id + gender_male + lunch_nonfree + ethnicity_amindian + ethnicity_asian + ethnicity_cauc + ethnicity_hispanic + ethnicity_other + class_type_regular_aide + class_type_small + school_rural + school_suburban + school_urban + degree_master + degree_masterplus + degree_specialist + ladder_apprentice + ladder_level1 + ladder_level2 + ladder_level3 + ladder_probation + t_ethnicity_cauc
                            OLS Regression Results                            
Dep. Variable:             score_math   R-squared:                       0.098
Model:                            OLS   Adj. R-squared:                  0.093
Method:                 Least Squares   F-statistic:                     16.92
Date:                Mon, 27 Oct 2025   Prob (F-statistic):           4.31e-67
Time:                        00:56:25   Log-Likelihood:                -19642.
No. Observations:                3749   AIC:                         3.933e+04
Df Residual